src/lib.rs

```diff
@@ -3466,6 +3467,255 @@ fn digraph_complement(
     Ok(complement_graph)
 }
 
+#[allow(clippy::too_many_arguments)]
+fn _spring_layout<Ty>(
+    py: Python,
+    graph: &StableGraph<PyObject, PyObject, Ty>,
+    pos: Option<HashMap<usize, layout::Point>>,
+    fixed: Option<HashSet<usize>>,
+    k: Option<f64>,
+    repulsive_exponent: Option<i32>,
+    adaptive_cooling: Option<bool>,
+    num_iter: Option<usize>,
+    tol: Option<f64>,
+    weight_fn: Option<PyObject>,
+    default_weight: f64,
+    scale: Option<f64>,
+    center: Option<layout::Point>,
+    seed: Option<u64>,
+) -> PyResult<Pos2DMapping>
+where
+    Ty: EdgeType,
+{
+    if fixed.is_some() && pos.is_none() {
+        return Err(PyValueError::new_err("`fixed` specified but `pos` not."));
+    }
+
+    let mut rng: Pcg64 = match seed {
+        Some(seed) => Pcg64::seed_from_u64(seed),
+        None => Pcg64::from_entropy(),
+    };
+
+    let dist = Uniform::new(0.0, 1.0);
+
+    let pos = pos.unwrap_or_default();
+    let mut vpos: Vec<layout::Point> = (0..graph.node_bound())
+        .map(|_| [dist.sample(&mut rng), dist.sample(&mut rng)])
+        .collect();
+    for (n, p) in pos.into_iter() {
+        vpos[n] = p;
+    }
+
+    let fixed = fixed.unwrap_or_default();
+    let k = k.unwrap_or(1.0 / (graph.node_count() as f64).sqrt());
+    let f_a = layout::AttractiveForce::new(k);
+    let f_r = layout::RepulsiveForce::new(k, repulsive_exponent.unwrap_or(2));
+
+    let num_iter = num_iter.unwrap_or(50);
+    let tol = tol.unwrap_or(1e-6);
+    let step = 0.1;
+
+    let mut weights: HashMap<(usize, usize), f64> = HashMap::default();
```

Comment:
```suggestion
    let mut weights: HashMap<(usize, usize), f64> = HashMap::with_capacity(graph.graph.edge_count());
```

Just to avoid multiple memory allocations.